Fix wrong columns getting processed on a CGC change

jimmygchen opened the pull request:
## Issue Addressed

This PR fixes a bug where wrong columns could get processed immediately after a CGC increase.

Scenario:
- The node's CGC increased due to additional validators attached to it (lets say from 10 to 11)
- The new CGC is advertised and new subnets are subscribed immediately, however the change won't be effective in the data availability check until the next epoch (See [this](https://github.com/sigp/lighthouse/blob/ab0e8870b4a35872574f5e33e9d13788c02ce250/beacon_node/beacon_chain/src/validator_custody.rs#L93-L99)). Data availability checker still only require 10 columns for the current epoch.
- During this time, data columns for the additional custody column (lets say column 11) may arrive via gossip as we're already subscribed to the topic, and it may be incorrectly used to satisfy the existing data availability requirement (10 columns), and result in this additional column (instead of a required one) getting persisted, resulting in database inconsistency.

## Proposed fix

- ~~In the `handle_gossip` function, we compute `should_process` for each data column that arrived, and pass it to the beacon processor - this column will get verified, propagated, but NOT processed / imported.~~ I've move this logic to `DataAvailabilityChecker` for consistency with RPC and better separation of concerns.
- This requires us to know the sampling columns for the current epoch, which is different to `globals.sampling_columns` (already updated to the new CGC). 
- To avoid having to maintain multiple column lists for current and next epoch, we compute the full ordered set of custody group count on startup, and this would allow us to quickly retrieve custody columns for both current and next epoch without having to recompute or maintain multiple lists.

The trade off here is we'd have to initialise the `CustodyContext` with the results from `get_custody_groups` _after_ network is initialised, which isn't ideal but will allow us to simplify and clean things up a lot by keeping custody related logic in `CustodyContext`. With this approach, we could potentially remove `network_globals.sampling_columns` too - I haven't done it in this PR though, wanted to see how this approach look like and get some feedback first.

The other options considered:
- Load `node_id` before `BeaconChain` is constructed, and have beacon chain also compute the custody columns. I think this leaks networking types into the `BeaconChain`, so IMO isn't great for separation of concerns and also introduces a circular dependency which will bite us later.
- Having network handle all the custody column related logic and only pass relevant data to the beacon processor. This approach splits custody logic between network and beacon chain - i believe the low cohesivsion sacrifices readability and maintainability.

jimmygchen committed:
Don't process extra custody groups that aren't effective yet but still verify and propagate them.

jimmygchen commented:
I've labeled it as `do-not-merge` as this still requires test fixes and clean up on docs, but the idea is there and I'd like to get some early feedback.

jimmygchen committed:
Move custody column filtering logic to `DataAvailabilityChecker` for consistency with RPC and better separation of concerns

jimmygchen commented on beacon_node/client/src/builder.rs:
```diff
@@ -787,6 +790,20 @@ where
+        .map_err(|e| format!("Failed to compute custody groups: {:?}", e))?;
+    chain
+        .data_availability_checker
+        .custody_context()
```
TODO: we can probably remove the `clone` in `custody_context()`

pawanjay176 commented on beacon_node/beacon_chain/src/validator_custody.rs:
```diff
@@ -170,7 +184,30 @@ impl CustodyContext {
+        all_custody_groups_ordered: Vec<CustodyIndex>,
+        spec: &ChainSpec,
+    ) -> Result<(), String> {
+        let mut ordered_custody_groups = vec![];
```
I don't understand this.
We basically want to store an ordering of data columns for a given `NodeId`.
Why are we storing columns for every custody index here? Can't we just compute the ordering for the node id and take a slice [..cgc] for the required cgc value?

jimmygchen replied:
yeah, this is due to custody groups decoupling - because number of columns could be different to custody groups. The alternative is storing all columns instead of groups, but it may not be as ergonomic because all the configs deal with groups (`sample_per_slot`,  `custody_requirement` etc)

`cgc` doesn't directly translate to columns either, we'd need to get the `sampling_size` (which also uses custody groups as unit), and then convert it to number of columns.

I think it's also doable though - so we'd
- store an ordered list of all columns
- when CGC changes, compute the `sampling_size` and then convert it to `sampling_size_columns`
- and take a slice of all columns using `sampling_size_columns`

The reason i did this was to potentially also store subnets in the `CustodyGroup` struct - it seems like a more logical grouping, and potentially remove `globals.sampling_subnets` and keep all custody related functions and values in `CustodyContext` - I think this is not necessary though, was just an idea that i had earlier, keeping networking logic in network is probably better.

pawanjay176 replied:
aghhh i hate the unnecessary abstraction with custody groups. I'm willing to bet that we will never change the columns per custody group and this will just be a painful tech debt forever 😭 

I also prefer having the columns to sample in one place to avoid syncing up everything if possible. 
I really like the approach of storing an ordered list of columns that you described. The computation is constant time so shouldn't be too expensive

jimmygchen replied:
yeah the custody groups abstraction complicates a lot of things, including the tests.

I've updated it to store an ordered list of columns, i think it seems to work out quite well, and also managed to get rid of `network_globals.sampling_columns`, which allowed us to remove `globals` from a bunch of places 🎉

pawanjay176 reviewed (commented):
I like the direction of this, its less invasive than I thought originally

jimmygchen committed:
Refactor custody context to store list of ordered columns instead.

jimmygchen committed:
Add tests and refactor.

jimmygchen committed:
Remove `sampling_columns` from `NetworkGlobals`.

jimmygchen commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -800,3 +814,209 @@ impl<E: EthSpec> MaybeAvailableBlock<E> {
+    /// Test to verify any extra RPC columns received that are not part of the "effective" CGC for
+    /// the slot are excluded from import.
+    #[test]
+    fn should_exclude_rpc_columns_not_required_for_sampling() {
```
Writing these tests make me want to use mocks again :sweat_smile: 

Without mocking, it requires understanding of the internals and how its dependencies work (in this case `custody_context`). The tests are just polluted with setting up the custody context. Whats worse is when the dependency behaviour changes, all these tests will break and require update.

jimmygchen replied:
I'll make another attempt with mocks tomorrow to compare.

jimmygchen replied:
Changed my mind, doesn't seem to be worth the effort now since the tests have already been written and doesn't seem _too_ bad now. We could reconsider this if maintenance becomes difficult.

jimmygchen committed:
Merge remote-tracking branch 'origin/unstable' into fix-extra-columns-processed

# Conflicts:
#	beacon_node/beacon_chain/src/fetch_blobs/tests.rs
#	beacon_node/http_api/src/publish_blocks.rs

jimmygchen committed:
Fix tests.

jimmygchen commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -233,15 +233,24 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
+        // a new epoch.
+        let sampling_columns = self
+            .custody_context
+            .sampling_columns_for_slot(Some(slot), &self.spec);
```
If we move this to the `overflow_lru_cache`, we could avoid the below altogether and make the `num_of_data_columns_to_sample` function private.

https://github.com/sigp/lighthouse/blob/ba02fe1da09f3ac28f6d6c03553de106e9afdb64/beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs#L530-L532

jimmygchen replied:
Rethinking about this again, feel like this is a micro-refactor that's not worth it, i think performance-wise this is slightly better than using `contains` to check every columns. Will leave it as it is.

pawanjay176 commented on consensus/types/src/data_column_custody_group.rs:
```diff
@@ -25,13 +24,32 @@ pub fn get_custody_groups(
     custody_group_count: u64,
     spec: &ChainSpec,
 ) -> Result<HashSet<CustodyIndex>, DataColumnCustodyGroupError> {
+    get_custody_groups_ordered(raw_node_id, custody_group_count, spec)
```
I'm wondering if the hashset is required anymore? why don't we make the default return the ordered one according to the spec?

jimmygchen replied:
As discussed last week, I  had considered it but the diff is quite big so I'll do this in the next PR!

pawanjay176 commented on beacon_node/network/src/sync/network_context.rs:
```diff
@@ -608,12 +613,7 @@ impl<T: BeaconChainTypes> SyncNetworkContext<T> {
-                        .iter()
-                        .copied()
-                        .collect(),
+                    self.chain.sampling_columns_for_slot(None).to_vec(),
```
don't we know the slot values here?

jimmygchen replied:
same as above

jimmygchen replied:
added

pawanjay176 commented on beacon_node/network/src/sync/network_context.rs:
```diff
@@ -551,7 +551,12 @@ impl<T: BeaconChainTypes> SyncNetworkContext<T> {
-                let column_indexes = self.network_globals().sampling_columns();
+                let column_indexes = self
+                    .chain
+                    .sampling_columns_for_slot(None)
```
don't we know the slot value here?

jimmygchen replied:
CGC can't change during sync so it's fine to use the slot I think, I was being a bit conservative and was preferring to download more and discard them in DA checker.

jimmygchen replied:
We discussed this last week - CGC doesn't change while syncing, so there should be no differences here and a bit more conservative to request the CGC at head to avoid batches failing if CGC is updated during sync (not possible right now). This is also the current behaviour. Let me know if you prefer to be explicit here, I also don't think it's a bad idea.

pawanjay176 replied:
this is relevant for backfill too though right? using the cgc at head could result in different backfill batches containing different number of columns based on the custody changes at head. 
Backfilling columns would have to account for different number of columns stored which isn't _that_ big a problem but its still a bit weird.

jimmygchen replied:
Yeah great catch, I hadn’t thought about backfill. Will fix this 🙏

jimmygchen replied:
Added epoch

pawanjay176 commented on beacon_node/beacon_chain/src/validator_custody.rs:
```diff
@@ -170,7 +177,35 @@ impl CustodyContext {
+        for custody_index in all_custody_groups_ordered {
+            let columns = compute_columns_for_custody_group(custody_index, spec)
+                .map_err(|e| format!("Failed to compute columns for custody group {e:?}"))?;
+            ordered_custody_groups.extend(columns);
```
this is `ordered_custody_columns` now no?

jimmygchen replied:
yes good catch

jimmygchen replied:
Fixed

pawanjay176 commented on beacon_node/beacon_chain/src/validator_custody.rs:
```diff
@@ -276,11 +311,37 @@ impl CustodyContext {
+    /// A slice of ordered column indices that should be sampled for this slot based on the node's custody configuration
+    pub fn sampling_columns_for_slot(
+        &self,
+        slot: Option<Slot>,
```
why not `sample_columns_for_epoch` considering we only switch the cgc at epoch boundaries? It'll be easier at sync too that way

pawanjay176 replied:
No need to have the EthSpec in the struct too that way

jimmygchen replied:
yeah I had considered it but thought slot would be more UX friendly for the caller - most of the time other than range sync we'd be using slots, so we'd have to do conversion at caller site.

`register_validator` also requires `EthSpec` though, feels like its safer and cleaner to have the generic at the struct level  for consistency (but maybe doenst matter because we only use one `EthSpec` anyway)

pawanjay176 replied:
I just felt that logically, it might be better to always have the assumption that all slots in an epoch have the same cgc. Having this function per slot seems like a potential footgun to me later on where we could refactor to different values at different slots within an epoch. But i'm fine with it either way

jimmygchen replied:
fair point, I'll update it! thanks

jimmygchen replied:
Updated the function to accept a mandatory epoch.

pawanjay176 reviewed (commented):
Looking good

jimmygchen committed:
Fix up variable name

jimmygchen commented:
Hey @pawanjay176 

Thanks for the review, I've addressed most of your comments and left responses for a few. I don't have strong opinion on those though, happy to make the changes, but thought I'd mention the thoughts behind them.

Let me know what you think!

pawanjay176 commented:
This is looking good, i'm testing this on the devnet with changing validator counts.

Update: testing this merged with my backfill branch and it works great 🎉  The only small thing is that it downloads different number of columns for different batches in backfill based on the custody count at head like i mentioned here https://github.com/sigp/lighthouse/pull/7792#discussion_r2252487293.

jimmygchen committed:
Require epoch instead of slot when checking sampling columns / custody groups.

jimmygchen commented:
Thanks @pawanjay176, I've addressed your review comments 🙏

jimmygchen committed:
Fix build.

jimmygchen committed:
Fix build.

jimmygchen committed:
Merge branch 'unstable' into fix-extra-columns-processed

pawanjay176 reviewed (approved):
LGTM. Retested it with the changes and works great.
